### DISCUSSION TITLE
Wanna learn to use ADKernelSUPG for application development

### DISCUSSION THREAD

**Discussion by Raven-pro** [idaholab/moose#22341] [Q&A General] [answered]
Hello everyone!

I am using MOOSE to build an application to simulate the transient single-phase flow using three basic equations and several state equations, and now it comes to the final part.

In the transient simulation, the oscillation happens near the boundary, leading to not converging at certain time step. So I am trying to use some stablization methods like SUPG used in the NS module.

Then I look into the NS kernels, finding that there is a basic kernel called ADKernelSUPG which is the base class of many ns kernels, so I am trying to make use of the Kernel to compute the stablization terms. 

However, I cannot find tutorial on how to use this kernel for development on the offical website. Could you help me with learning to use ADKernelSUPG and Kernels like that? Like examples or docs and things like that.

Thank you so much!

**Comment by GiudGiud** [comment]
Hello

I'm afraid the actual development of the NS module, especially the INSAD implementation, is not documented beyond the doxygen
https://mooseframework.inl.gov/docs/doxygen/moose/classADKernelSUPGTempl.html
and the modules FE description
https://mooseframework.inl.gov/modules/navier_stokes/cgfe.html

@lindsayad do you know of any more SUPG documentation out there?

Before you attempt to use SUPG, have you made sure your implementation is correct? By doing verification & validation on simple cases? see this post https://github.com/idaholab/moose/discussions/22344 for details on verification

Guillaume

**Comment by lindsayad** [comment, marked as answer]
The only class in MOOSE that inherits directly from `ADKernelSUPG` (the scalar field form) is [INSADEnergySUPG](https://mooseframework.inl.gov/source/kernels/INSADEnergySUPG.html). I think that doc page, although short, is helpful. Essentially you need to do 3 things:

- provide a material property describing `tau`
- provide a vector variable `velocity`
- implement `precomputeQpStrongResidual`. You could do this on a strong term basis, e.g. if you have three terms in your PDE, create three kernels that inherit from `ADKernelSUPG`. Or you could do something like what is done in `INSADEnergySUPG` where only one kernel is used to gather the contributions from all terms in the PDE (those contributions are held in a material property `_temperature_strong_residual`)

Strong term/strong form just means your PDE before you convert it to the weak form for finite element discretization.

**Reply by Raven-pro** [reply]
Thanks for your reply!
I will try on it  :)
